python/src/iceberg/io/pyarrow.py

```diff
@@ -0,0 +1,213 @@
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""FileIO implementation for reading and writing table files that uses pyarrow.fs
+
+This file contains a FileIO implementation that relies on the filesystem interface provided
+by PyArrow. It relies on PyArrow's `from_uri` method that infers the correct filesytem
+type to use. Theoretically, this allows the supported storage types to grow naturally
+with the pyarrow library.
+"""
+
+from typing import Union
+from urllib.parse import ParseResult, urlparse
+
+from pyarrow import NativeFile
+from pyarrow.fs import FileSystem, FileType
+
+from iceberg.io.base import FileIO, InputFile, InputStream, OutputFile, OutputStream
+
+
+class PyArrowInputFile(InputFile):
+    """An InputFile implementation that uses a pyarrow filesystem to generate pyarrow.lib.NativeFile instances for reading
+
+    Args:
+        location(str): The URI to the file
+
+    Attributes:
+        location(str): The URI to the file
+        parsed_location(urllib.parse.ParseResult): The parsed location URI
+        exists(bool): Whether the file exists or not
+
+    Examples:
+        >>> from iceberg.io.pyarrow import PyArrowInputFile
+        >>> input_file = PyArrowInputFile("s3://foo/bar.txt")
+        >>> file_content = input_file.open().read()
+    """
+
+    def __init__(self, location: str):
+        parsed_location = urlparse(location)  # Create a ParseResult from the uri
+
+        if parsed_location.scheme not in (
+            "file",
+            "mock",
+            "s3fs",
+            "hdfs",
+            "viewfs",
+        ):  # Validate that a uri is provided with a scheme of `file`
+            raise ValueError("PyArrowInputFile location must have a scheme of `file`, `mock`, `s3fs`, `hdfs`, or `viewfs`")
+
+        super().__init__(location=location)
+        self._parsed_location = parsed_location
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        filesytem, path = FileSystem.from_uri(self.location)  # Infer the proper filesystem
+        file = filesytem.open_input_file(path)
+        return file.size()
+
+    @property
+    def parsed_location(self) -> ParseResult:
+        """The parsed location
+
+        Returns:
+            ParseResult: The parsed results which has attributes `scheme`, `netloc`, `path`,
+            `params`, `query`, and `fragments`.
+        """
+        return self._parsed_location
+
+    @property
+    def exists(self) -> bool:
+        """Checks whether the file exists"""
+        filesytem, path = FileSystem.from_uri(self.location)  # Infer the proper filesystem
+        file_info = filesytem.get_file_info(path)
+        return False if file_info.type == FileType.NotFound else True
+
+    def open(self) -> NativeFile:
```

Comment:
Yes, I would thinking of `-> InputStream`, I haven't coded much in python since typehints have come about, but it feels like generally for method overrides its not clear if you want specify a more specific type then the method that is getting overridden suggests.

Reply:
I just did a quick test and mypy will indeed check the overrides against the protocol. This is the right flexibility since if for example pyarrow one day decided to return `CustomFooFile` instead of `NativeFile`, mypy would raise the following **_only_** if it's missing the methods defined in the protocol:
```
Return type "CustomFooFile" of "open" incompatible with return type "InputStream" in supertype "InputFile"
```
So it feels like what we want--to only enforce adherence to the protocols.

Reply:
Updated the return types to be `InputStream` and `OutputStream`. 👍